### PR TITLE
Go backend

### LINKED ISSUE #15: Authentication
- [x] Admins
    - Are able to see all information about users + products
    - Are able to edit user and product information
- [x] Users
    - Are only able to see products and their own balance

### LINKED ISSUE #26: Routes
Properly route every documented route.

### PR DESCRIPTION
**Pull request by TadMaj**
Finalized all routing and continueing to finalize basic functionality

### CODE REVIEW

**Review by EStroiu — APPROVED** [review]

File: knt-backend/tools/sqltest/main.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
+package main
```
Comment (trixmoe) [inline knt-backend/tools/sqltest/main.go:1]: Explain what this is for. Apparently for debugging.

File: knt-backend/run-docker.sh [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+# Build docker image
+docker build -t docker-knt-backend:multistage -f Dockerfile.multistage .
+
+# Stop and remove docker container
+docker stop knt-backend
+docker rm knt-backend
+
+# Run docker image
+docker run -p 5000:5000 --name knt-backend -v "$(pwd)"/database:/database -v "$(pwd)"/logs:/logs docker-knt-backend:multistage
```
Comment (trixmoe) [inline knt-backend/run-docker.sh:12]: Building/Running should be separate.
Also, you shouldn't have to stop/rm, I'm sure there's a cleaner way around this.
On top of that, you should use docker-compose and not just a run command.

File: knt-backend/internal/kntrouter/user_requests.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+	err := decoder.Decode(&body)
+	if err != nil {
+		logger.Error("Unable to decode body: ", err.Error())
+		http.Error(w, err.Error(), http.StatusUnprocessableEntity)
+		return
+	}
+
+	_, err = kntdb.UpdateUser(body)
+	if err != nil {
+		logger.Error("Unable to decode body: ", err.Error())
```
Comment (trixmoe) [inline knt-backend/internal/kntrouter/user_requests.go:143]: Uable to decode the body _twice_? I think you forgot to change the log message?

File: knt-backend/internal/kntrouter/user_requests.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+		logger.Error("Unable to update user balance: ", err.Error(), " vunetid: ", format.VunetId)
+		http.Error(w, err.Error(), http.StatusUnprocessableEntity)
+		return
+	}
+	w.WriteHeader(http.StatusNoContent)
+}
+
+func getTransactions(w http.ResponseWriter, r *http.Request) {
+	pp := r.URL.Query().Get("pp")
+	p := r.URL.Query().Get("p")
```
Comment (trixmoe) [inline knt-backend/internal/kntrouter/user_requests.go:187]: If you're leaving it with `p` and `pp`, at least make it clear what it means.
Its not readable for non-php developers.

File: docs/knt-backend-doc.yml [hunk 5]
```diff
@@ -64,19 +64,33 @@ paths:
       description: 'Edit product'
```
Comment (trixmoe) [inline docs/knt-backend-doc.yml:64]: See the messages I sent you about the documentation. There's stuff that doesn't align up.

File: knt-backend/Dockerfile.multistage [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+COPY ./go.mod ./go.sum ./main.go ./
+COPY ./internal ./internal
+
+RUN go mod download
+
+RUN GOOS=linux go build -o /knt .
+
+# Deploy the application binary into a lean image
+FROM gcr.io/distroless/base-debian11 AS build-release-stage
```
Comment (trixmoe) [inline knt-backend/Dockerfile.multistage:13]: This feels cursed, maybe we should check at using Alpine before we deploy it.

File: knt-backend/config/kntconfig.yml [hunk 7]
```diff
@@ -24,8 +26,8 @@ routes:
 
   getUserAdmin: "/users/{userId}"
   createNewUser: "/users"
-  updateUser: "/users/{userId}"
-  updateUserMoney: "/users/{userId}/balance"
+  updateUser: "/users"
+  updateUserMoney: "/users/balance"
```
Comment (trixmoe) [inline knt-backend/config/kntconfig.yml:30]: Just make sure this lines up with the doc.

File: knt-backend/internal/kntrouter/middleware.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+	r.Use(userMiddleware)
+}
+
+// This sets the cors for all requests, this should be edited in the config
+func setCors(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Access-Control-Allow-Origin", viper.GetString("corsAllowOrigin"))
+		w.Header().Set("Access-Control-Allow-Credentials", viper.GetString("corsAllowCredentials"))
+		w.Header().Set("Access-Control-Allow-Methods", viper.GetString("corsAllowMethods"))
+		w.Header().Set("Access-Control-Allow-Headers", viper.GetString("corsAllowHeaders"))
```
Comment (trixmoe) [inline knt-backend/internal/kntrouter/middleware.go:39]: Don't forget to fix this before deploying it...

File: knt-backend/internal/kntdb/user_database.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	}
+	old.Visibility = new.Visibility
+	if new.Password != "" {
+		old.Password = ShaHashing(new.Password)
+	}
+	return old
+}
+
+func GetPopulatedTransactions(pp int, p int) ([]TransactionDetails, error) {
+	basicTrans, err := getBasicTransactions(pp, p)
```
Comment (trixmoe) [inline knt-backend/internal/kntdb/user_database.go:72]: again, p and pp are not clear.

File: knt-backend/internal/kntdb/receipt.go [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+		}
+	}
+
+	return nil
+}
+
+func ensureTaxTableExists(transaction *sql.Tx, year int) error {
+	_, err := addToTransaction(transaction, "CREATE TABLE IF NOT EXISTS tax (id INTEGER PRIMARY KEY AUTOINCREMENT, product_id INT, amount INT, totalCost INT, year INT)")
+	return err
+}
```
Comment (trixmoe) [inline knt-backend/internal/kntdb/receipt.go:171]: This is **_extremely_** cursed. I'm genuinely **not** willing to accept this.
You are **computing** the total taxes (good), but you are storing it back in the database, as if it were a fixed value, which it IS NOT.
On top of that, you are not even re-creating the table from scratch. So if you run this again, you end up with what you had before + the taxes again?! So like 2x taxes?!

This is cursed beyond belief. You should avoid duplicate data! Just use an SQL query to get the SUM (amount) of all the purchases of a certain item type and then multiply that by the tax percentage?
This is the job of a query, not of stored data.

Also, do you even commit the transaction? For some reasons I have doubts this even works.

File: knt-backend/internal/kntdb/product_database.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+func CreateNewProduct(product Product) (int64, error) {
+	return commitTransaction(
+		"insert into product (price, name, visibility, taxcategory) VALUES (?, ?, ?, ?)",
+		product.Price, product.Name, product.Visibility, product.TaxCategory)
+}
+
+func UpdateProduct(product Product) (int64, error) {
+	if product.Id == 0 {
+		return 0, errors.New("invalid user")
```
Comment (trixmoe) [inline knt-backend/internal/kntdb/product_database.go:31]: invalid user? not product? i feel like a lot of it is copy-pasted :/

**Review by trixmoe — COMMENTED** [review]
A lot of it is "eh" but alright, some of it is really not great though, for instance the taxes.

At this point, I'm happy to see that we are slowly getting close to a working, more modern system, but I'm also starting to see some of the flaws of this approach.
I believe all of this, and most particularly the ongoing documentation efforts, will be extremely useful for the years to come, as to make the whole system more clear.

I'm hoping we can get more documentation for various projects, as to clear up a lot of the incomprehensibility among STORM (CHIP) projects.

**Pull request merged**